vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,666 @@
+    {
+      vec.push_back({(char)action_index, (char)ccb_id_namespace});
+    }
+  }
+
+  for (auto& shared_index : shared->indices)
+  {
+    vec.push_back({(char)shared_index, (char)ccb_id_namespace});
+  }
+}
+
+// build a cb example from the ccb example
+template <bool is_learn>
+void build_cb_example(multi_ex& cb_ex, example* slot, ccb& data)
+{
+  bool slot_has_label = slot->l.conditional_contextual_bandit.outcome != nullptr;
+
+  // Merge the slot features with the shared example and set it in the cb multi-example
+  // TODO is it imporant for total_sum_feat_sq and num_features to be correct at this point?
+  inject_slot_features(data.shared, slot);
+  cb_ex.push_back(data.shared);
+
+  // For V0, include list is not supported
+  // retrieve the action index whitelist (if the list is empty, then all actions are white-listed)
+  // data.include_list.clear();
+  // for (uint32_t included_action_id : slot->l.conditional_contextual_bandit.explicit_included_actions)
+  //   data.include_list.insert(included_action_id);
+
+  // set the available actions in the cb multi-example
+  uint32_t index = 0;
+  data.origin_index.clear();
+  data.origin_index.resize(data.actions.size(), 0);
+  for (size_t i = 0; i < data.actions.size(); i++)
+  {
+    // For V0, include list is not supported
+    // filter actions that are not explicitly included
+    // if (!data.include_list.empty() && data.include_list.find((uint32_t)i) == data.include_list.end())
+    //   continue;
+
+    // filter actions chosen by previous slots
+    if (data.exclude_list[i])
+      continue;
+
+    // select the action
+    cb_ex.push_back(data.actions[i]);
+
+    // save the original index from the root multi-example
+    data.origin_index[index++] = (uint32_t)i;
+
+    // remember the index of the chosen action
+    if (is_learn && slot_has_label && i == slot->l.conditional_contextual_bandit.outcome->probabilities[0].action)
+    {
+      // This is used to remove the label later.
+      data.action_with_label = (uint32_t)i;
+      attach_label_to_example(index, data.actions[i], slot->l.conditional_contextual_bandit.outcome, data);
+    }
+  }
+
+  // Must reset this in case the pooled example has stale data here.
+  data.shared->pred.a_s = data.action_score_pool.get_object();
+
+  // Tag can be used for specifying the sampling seed per slot. For it to be used it must be inserted into the shared
+  // example.
+  std::swap(data.shared->tag, slot->tag);
+}
+
+// iterate over slots contained in the multi-example, and for each slot, build a cb example and perform a
+// cb_explore_adf call.
+template <bool is_learn>
+void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
+{
+  clear_all(data);
+  split_multi_example(examples, data);  // split shared, actions and slots
+
+#ifndef NDEBUG
+  sanity_checks<is_learn>(data);
+#endif
+
+  // Stash the CCB labels before rewriting them.
+  data.stored_labels.clear();
+  for (auto ex : examples)
```

Comment:
>for [](start = 2, length = 3)

Move this into split_multi_examples?  You are iterating the example collection there anyway.  Might require renaming split_multi_example()